### ISSUE TITLE
Not matching on colorized clippy lints, is this expected?

### ISSUE THREAD

**Issue by obi1kenobi** [r7kamura/rust-problem-matchers#8]
Thanks for this awesome GitHub Action!

I started using it in my project somewhat recently, and I was wondering if the action is expected to match on clippy lints as well as rustc lints and test errors.

For example: I added some dead code on purpose in [this PR](https://github.com/obi1kenobi/trustfall/pull/354) and the lint job (which [uses this action](https://github.com/obi1kenobi/trustfall/blob/main/.github/workflows/ci.yml#L156-L167)) [failed as expected](https://github.com/obi1kenobi/trustfall/actions/runs/5490852131/jobs/10006790730?pr=354). However, I didn't see any problem matcher output, so either I have misconfigured something, or something else is going wrong.

This isn't terribly high-priority for me, so no rush. I appreciate any help you're able to offer!

**Comment by r7kamura** [comment]
Thanks for reporting 👍

Maybe the following regex patterns should match that error, but it looks like something is happening that is not as intended.

https://github.com/r7kamura/rust-problem-matchers/blob/d58b70c4a13c4866d96436315da451d8106f8f08/.github/matchers.json#L3-L19

Looking at the logs, it seems that the output is colored by ANSI escape sequences, unlike when I implemented it, and this type of problem matchers doesn't seem to be able to handle colored output.

Could you try removing the color from the output in some way?

**Comment by obi1kenobi** [comment]
The color is from the env variable `CARGO_TERM_COLOR=always` set here:
https://github.com/obi1kenobi/trustfall/blob/main/.github/workflows/ci.yml#L15

Removing the color does indeed fix the problem: https://github.com/obi1kenobi/trustfall/pull/354/files

I'm assuming it should be possible to update the regex to match on (and ignore) ANSI escape sequences. I found a regex pattern on StackOverflow which seems like it could be a solid starting point: https://stackoverflow.com/questions/14693701/how-can-i-remove-the-ansi-escape-sequences-from-a-string-in-python

Is this something where you'd be open to accepting a PR? I don't know exactly how GitHub Matchers work but I'm willing to work it out :)

**Comment by r7kamura** [comment]
> I'm assuming it should be possible to update the regex to match on (and ignore) ANSI escape sequences. 

Indeed. However, I am wondering if it should be handled that way, as I think I will have to maintain very complex regexes compared to the current simple ones... 🤔

**Comment by obi1kenobi** [comment]
Unfortunately I frequently find myself looking at raw output logs even when matchers work, so not colorizing the output wouldn't work for me since I find the uncolored output too hard to read.

I understand the maintainability concern. If there's something I can do to help address it, I'd be open to it.

AFAICT the color only appears on leading words on a line, so I don't think the regex change would be too significant. I'm happy to make a PR and test it.

If you don't think the maintainability concern of colorized output can be addressed, I totally understand and I don't mind being told to write my own matchers :)

**Comment by antifuchs** [comment]
Huh, I'm seeing a similar issue but _with_ `CARGO_TERM_COLOR=never`: Using stable clippy, this matcher does not find the line number message ([see](https://github.com/boinkor-net/ci-baseline-rust/actions/runs/5777139499/job/15656979983#step:5:11)). Looks like clippy emits a `##[error]` prefix to my messages, which seems to not match the second line in the rust-compiler matcher. Checking the raw logs, it looks like the extra context line is this:

```
2023-08-06T14:23:50.3949805Z ##[error] --> src/main.rs:3:12
```

...but maybe that's something that github actions inserts? I've played with the regex in my fork at https://github.com/antifuchs/rust-problem-matchers/tree/clippy-regex, but couldn't find a regex that gives the problem matcher the ability to extract the file/line number/column from the clippy message.

**Comment by antifuchs** [comment]
Oops, my bad: The files that clippy errors about are in a directory other than the project root, so github can't find where the actual file path is, from the relative path clippy prints. Don't think clippy even prints the current directory anywhere, so it's probably impossible to get good links on non-root directory files.

**Comment by obi1kenobi** [comment]
> > I'm assuming it should be possible to update the regex to match on (and ignore) ANSI escape sequences.
> 
> Indeed. However, I am wondering if it should be handled that way, as I think I will have to maintain very complex regexes compared to the current simple ones... 🤔

I had a brief look at ANSI escape sequences, and the colorization ones cargo uses seem decently straightforward: `\x1B\[[0-9;]*[a-zA-Z]` is what `cargo-semver-checks` uses to match on and eliminate ANSI sequences [here](https://github.com/obi1kenobi/cargo-semver-checks/blob/main/.github/workflows/ci.yml#L1317-L1318).

Would adding a pattern like that still a be maintainability concern on your end?

I'd love to contribute a PR for it if you're open to it. If not, obviously no hard feelings at all — with your permission (since I didn't see a LICENSE file), I don't mind keeping my own fork.

**Comment by r7kamura** [comment]
Thank you @obi1kenobi. I would be happy to have you send me a pull request, as I think the pattern as simple as that would not be a problem 👍 

In addition, I added a license file for the MIT license at https://github.com/r7kamura/rust-problem-matchers/pull/10.

**Comment by obi1kenobi** [comment]
`rustc` and `clippy` add the escape sequences in a few different places so I had to put the same pattern in a few places as well. I tried to make everything as clear as possible by intentionally keeping the ANSI escape sequence matcher identical in all the places, and I've explained all the changes in detail in the commit message / PR summary as well.

Thanks for being open to a PR for this, I'm excited to use it! 🎉
